File: src/med_exam_toolkit/parsers/ahuyikao.py
from __future__ import annotations
from med_exam_toolkit.models import Question, SubQuestion
from med_exam_toolkit.parsers import register
from med_exam_toolkit.parsers.base import BaseParser


@register("ahuyikao")
class AhuyikaoParser(BaseParser):

    def can_handle(self, raw: dict) -> bool:
        return raw.get("pkg", "") == "ahuyikao.com"

    def parse(self, raw: dict) -> Question:
        mode = raw.get("mode", "")
        q = Question(
            name=raw.get("name", ""),
            pkg=raw.get("pkg", ""),
            cls=raw.get("cls", ""),
            unit=raw.get("unit", ""),
            mode=mode,
            raw=raw,
        )

        if "B" in mode and "型题" in mode:
            q.shared_options = raw.get("shared_options", [])
            q.discuss = raw.get("discuss", "")
            for sq in raw.get("sub_questions", []):
                q.sub_questions.append(SubQuestion(
                    text=sq.get("test", ""),
                    options=q.shared_options,
                    answer=sq.get("answer", ""),
                    rate=sq.get("rate", ""),
                    error_prone=sq.get("error_prone", ""),
                    discuss=sq.get("discuss", ""),
                ))

        elif "A3" in mode or "A4" in mode:
            q.stem = raw.get("stem", "")
            for sq in raw.get("sub_questions", []):
                q.sub_questions.append(SubQuestion(
                    text=sq.get("test", ""),
                    options=sq.get("option", []),
                    answer=sq.get("answer", ""),
                    rate=sq.get("rate", ""),
                    error_prone=sq.get("error_prone", ""),
                    discuss=sq.get("discuss", ""),
                ))

        else:
            # A1 / A2 型题
            q.sub_questions.append(SubQuestion(
                text=raw.get("test", ""),
                options=raw.get("option", []),
                answer=raw.get("answer", ""),
                rate=raw.get("rate", ""),
                error_prone=raw.get("error_prone", ""),
                discuss=raw.get("discuss", ""),
            ))

        return q

File: internal/parsers/ahuyikao.go
package parsers

import (
	"github.com/zyh001/med-exam-kit/internal/models"
)

// AhuyikaoParser handles JSON from com.ahuxueshu / ahuyikao.com.
type AhuyikaoParser struct{}

func (p *AhuyikaoParser) CanHandle(raw map[string]any) bool {
	return str(raw["pkg"]) == "com.ahuxueshu"
}

func (p *AhuyikaoParser) Parse(raw map[string]any) (*models.Question, error) {
	mode := str(raw["mode"])
	q := &models.Question{
		Name: str(raw["name"]),
		Pkg:  str(raw["pkg"]),
		Cls:  str(raw["cls"]),
		Unit: str(raw["unit"]),
		Mode: mode,
		Raw:  raw,
	}

	switch {
	case contains(mode, "B") && contains(mode, "型题"):
		q.SharedOptions = strSlice(raw["shared_options"])
		q.Discuss = str(raw["discuss"])
		for _, item := range anySlice(raw["sub_questions"]) {
			sq := item.(map[string]any)
			q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
				Text:       str(sq["test"]),
				Options:    strSliceCopy(q.SharedOptions), // independent copy
				Answer:     str(sq["answer"]),
				Rate:       str(sq["rate"]),
				ErrorProne: str(sq["error_prone"]),
				Discuss:    str(sq["discuss"]),
			})
		}

	case contains(mode, "A3") || contains(mode, "A4"):
		q.Stem = str(raw["stem"])
		for _, item := range anySlice(raw["sub_questions"]) {
			sq := item.(map[string]any)
			q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
				Text:       str(sq["test"]),
				Options:    strSlice(sq["option"]),
				Answer:     str(sq["answer"]),
				Rate:       str(sq["rate"]),
				ErrorProne: str(sq["error_prone"]),
				Discuss:    str(sq["discuss"]),
			})
		}

	default: // A1 / A2
		q.SubQuestions = append(q.SubQuestions, models.SubQuestion{
			Text:       str(raw["test"]),
			Options:    strSlice(raw["option"]),
			Answer:     str(raw["answer"]),
			Rate:       str(raw["rate"]),
			ErrorProne: str(raw["error_prone"]),
			Discuss:    str(raw["discuss"]),
		})
	}
	return q, nil
}

// ── shared helpers (used by both parsers) ─────────────────────────────

func str(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func strSlice(v any) []string {
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, item := range arr {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func strSliceCopy(src []string) []string {
	c := make([]string, len(src))
	copy(c, src)
	return c
}

func anySlice(v any) []any {
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	return arr
}

func contains(s, sub string) bool {
	return len(sub) > 0 && len(s) >= len(sub) &&
		func() bool {
			for i := 0; i <= len(s)-len(sub); i++ {
				if s[i:i+len(sub)] == sub {
					return true
				}
			}
			return false
		}()
}
